### ISSUE TITLE
Tests fail because OpenShift container vanishes underneath

### ISSUE THREAD

**Issue by Lehtux** [zalando/zalenium#518]
Zalenium Image Version(s):  
3.11.0d
Selenium image:
3.11.0-p5

OS: 
OpenShift Enterprise

## Expected Behavior -
Tests will pass through normally.
## Actual Behavior -
Seems like Zalenium is closing containers even though a test case is registered there?
Getting this error from hub under "heavy load":
`15:10:11.675 [qtp565372776-2332] INFO  o.o.g.w.s.handler.RequestHandler - Error forwarding the new session Error forwarding the request unexpected end of stream on Connection{10.1.16.158:40049, proxy=DIRECT hostAddress=/10.1.16.158:40049 cipherSuite=none protocol=http/1.1}
org.openqa.grid.internal.exception.NewSessionException: Error forwarding the request unexpected end of stream on Connection{10.1.16.158:40049, proxy=DIRECT hostAddress=/10.1.16.158:40049 cipherSuite=none protocol=http/1.1}
	at org.openqa.grid.web.servlet.handler.RequestHandler.forwardNewSessionRequestAndUpdateRegistry(RequestHandler.java:94)
	at org.openqa.grid.web.servlet.handler.RequestHandler.process(RequestHandler.java:114)
	at org.openqa.grid.web.servlet.DriverServlet.process(DriverServlet.java:86)
	at org.openqa.grid.web.servlet.DriverServlet.doPost(DriverServlet.java:70)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.seleniumhq.jetty9.servlet.ServletHolder.handle(ServletHolder.java:860)
	at org.seleniumhq.jetty9.servlet.ServletHandler.doHandle(ServletHandler.java:535)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
	at org.seleniumhq.jetty9.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
	at org.seleniumhq.jetty9.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
	at org.seleniumhq.jetty9.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.seleniumhq.jetty9.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
	at org.seleniumhq.jetty9.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
	at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.seleniumhq.jetty9.server.Server.handle(Server.java:530)
	at org.seleniumhq.jetty9.server.HttpChannel.handle(HttpChannel.java:347)
	at org.seleniumhq.jetty9.server.HttpConnection.onFillable(HttpConnection.java:256)
	at org.seleniumhq.jetty9.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.seleniumhq.jetty9.io.FillInterest.fillable(FillInterest.java:102)
	at org.seleniumhq.jetty9.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
	at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
	at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.seleniumhq.jetty9.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
	at org.seleniumhq.jetty9.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
	at org.seleniumhq.jetty9.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: unexpected end of stream on Connection{10.1.16.158:40049, proxy=DIRECT hostAddress=/10.1.16.158:40049 cipherSuite=none protocol=http/1.1}
	at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:205)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at org.openqa.selenium.remote.internal.OkHttpClient.execute(OkHttpClient.java:101)
	at org.openqa.grid.internal.TestSession.sendRequestToNode(TestSession.java:430)
	at org.openqa.grid.internal.TestSession.forward(TestSession.java:237)
	at org.openqa.grid.web.servlet.handler.RequestHandler.forwardNewSessionRequestAndUpdateRegistry(RequestHandler.java:91)
	... 31 common frames omitted
Caused by: java.io.EOFException: \n not found: limit=0 content=…
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:227)
	at okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:212)
	at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189)
	... 50 common frames omitted
15:10:12.281 [Terminate Test Session int id: [ec2fe1d1-cb5e-4bed-bc2d-9ac71da8b9ee] ext id: [No external key was assigned] container: [taisto-40049-lrnh4]] INFO  d.z.e.z.c.k.KubernetesContainerClient - taisto-40049-lrnh4 
`

**Comment by diemol** [comment]
Hi @Lehtux,

Can you share some more details please? How many tests were you running in parallel, how many pods were running, and if possible your deployment yaml.

**Comment by peerster** [comment]
I think I ran into the same issue, or at least similiar.
We also run in openshift.

Checking the logs, my quick analys of the problem is:
1. Zalenium says it’s shutting down a node after test is finished (we only run one test per node)
2. A second later the same node as Zalenium said is shutting down is registering as a new node
3. Test gets sent angainst this registered node, which I guess is terminating.
4. We got one spawned container (besides the zalenium pod) that never terminates (zalenium-40034-fxfl7)

And the resulting error for the client is:
```cannot forward the request unexpected end of stream on Connection{10.128.7.46:40033, proxy=DIRECT hostAddress=/10.128.7.46:40033 cipherSuite=none protocol=http/1.1```

I've attached a large portion of the output we got from zalenium pod. I think the problems starts at `{"@timestamp":"2018-04-10T15:37:32.672+02:00"`
[zalenium-output.txt](https://github.com/zalando/zalenium/files/1895069/zalenium-output.txt)

**Comment by pearj** [comment]
Hi @peerster,

My colleague @dspasojevic has fixed the behaviour you're talking about in his restructure.  He observed the same thing, that nodes would re-register during shutdown sometimes.

I presume this because the nodes have a polling cycle, where they periodically re-register to the grid.  So if you wait for the restructure to land it _should_ be fixed.

**Comment by pearj** [comment]
I think he solved this, by marking the node as shutting down, but doesn't remove it from the registry until the container client can detect that the node is actually gone.

Otherwise, if you remove it from the registry before shutdown has actually finished, it might re-register but then be shutdown soon after.

**Comment by pearj** [comment]
@lehtux He’s currently fixing the tests. But I’m hoping he’ll push a PR in the next few days so that people can start trying it out.

**Comment by Lehtux** [comment]
@pearj awesome! Thanks!

**Comment by pearj** [comment]
@Lehtux First cut of code has just landed, if you're able to build it yourself you could try it out see #527

**Comment by Lehtux** [comment]
@pearj thanks! I'll give it a go.

**Comment by Lehtux** [comment]
@pearj @dspasojevic , having trouble of building this precise image, just stucks / spams here:
`INFO   Unregistering the node http://localhost:40000 because it's been down for 6341628 milliseconds`

**Comment by dspasojevic** [comment]
@Lehtux I've Ignore'd the two tests I've broken while I fix them. It should build for you now.

**Comment by peerster** [comment]
Also, you can build without the tests:
```
mvn clean package -DskipTests=true -Pbuild-docker-image
```

**Comment by Lehtux** [comment]
@dspasojevic 
Still getting the same kind of forward error:
```

13:05:19.823 [qtp1301987508-403] INFO  o.o.g.w.s.handler.RequestHandler - Got a request to create a new session: Capabilities {acceptInsecureCerts: true, browserName: firefox, platformName: linux}
--
  | 13:05:19.824 [qtp1301987508-403] INFO  d.z.e.z.registry.ZaleniumRegistry - Adding sessionRequest for {browserName=firefox, platformName=linux, acceptInsecureCerts=true}
  | 13:05:19.825 [Matcher thread] INFO  d.z.e.z.proxy.AutoStartProxySet - No proxy available for new session, starting new.
  | 13:05:19.854 [Matcher thread] INFO  d.z.e.z.p.DockeredSeleniumStarter - Created container [taisto-40000-8h672] with dimensions [(1920, 1080)] and tz [sun.util.calendar.ZoneInfo[id="Europe/Helsinki",offset=7200000,dstSavings=3600000,useDaylight=true,transitions=118,lastRule=java.util.SimpleTimeZone[id=Europe/Helsinki,offset=7200000,dstSavings=3600000,useDaylight=true,startYear=0,startMode=2,startMonth=2,startDay=-1,startDayOfWeek=1,startTime=3600000,startTimeMode=2,endMode=2,endMonth=9,endDay=-1,endDayOfWeek=1,endTime=3600000,endTimeMode=2]]].
  | 13:05:19.854 [Matcher thread] INFO  d.z.e.z.proxy.AutoStartProxySet - Created ContainerCreationStatus [isCreated=true, containerName=taisto-40000-8h672, nodePort=40000].
  | 13:05:19.952 [qtp1301987508-29] INFO  o.o.g.w.s.handler.RequestHandler - Error forwarding the new session null
  | java.lang.NullPointerException: null
  | at de.zalando.ep.zalenium.proxy.DockerSeleniumRemoteProxy.afterCommand(DockerSeleniumRemoteProxy.java:310)
  | at org.openqa.grid.internal.TestSession.forward(TestSession.java:281)
  | at org.openqa.grid.web.servlet.handler.RequestHandler.forwardNewSessionRequestAndUpdateRegistry(RequestHandler.java:91)
  | at org.openqa.grid.web.servlet.handler.RequestHandler.process(RequestHandler.java:114)
  | at org.openqa.grid.web.servlet.DriverServlet.process(DriverServlet.java:86)
  | at org.openqa.grid.web.servlet.DriverServlet.doPost(DriverServlet.java:70)
  | at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
  | at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
  | at org.seleniumhq.jetty9.servlet.ServletHolder.handle(ServletHolder.java:860)
  | at org.seleniumhq.jetty9.servlet.ServletHandler.doHandle(ServletHandler.java:535)
  | at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
  | at org.seleniumhq.jetty9.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
  | at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
  | at org.seleniumhq.jetty9.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
  | at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
  | at org.seleniumhq.jetty9.servlet.ServletHandler.doScope(ServletHandler.java:473)
  | at org.seleniumhq.jetty9.server.session.SessionHandler.doScope(SessionHandler.java:1564)
  | at org.seleniumhq.jetty9.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
  | at org.seleniumhq.jetty9.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
  | at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
  | at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
  | at org.seleniumhq.jetty9.server.Server.handle(Server.java:530)
  | at org.seleniumhq.jetty9.server.HttpChannel.handle(HttpChannel.java:347)
  | at org.seleniumhq.jetty9.server.HttpConnection.onFillable(HttpConnection.java:256)
  | at org.seleniumhq.jetty9.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
  | at org.seleniumhq.jetty9.io.FillInterest.fillable(FillInterest.java:102)
  | at org.seleniumhq.jetty9.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
  | at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
  | at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
  | at org.seleniumhq.jetty9.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
  | at org.seleniumhq.jetty9.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
  | at org.seleniumhq.jetty9.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
  | at org.seleniumhq.jetty9.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
  | at java.lang.Thread.run(Thread.java:748)

**Comment by dspasojevic** [comment]
@Lehtux can you tell me which test that is? The tests are passing in Travis at the moment.

**Comment by Lehtux** [comment]
@dspasojevic it's not coming from build process. This is the actual logs from Hub running on OpenShift.

**Comment by dspasojevic** [comment]
Hi @Lehtux are you getting that error using the latest from https://github.com/dspasojevic/zalenium ?

**Comment by Lehtux** [comment]
@dspasojevic ahh didn't see the new commits, I'll deploy again and check. Thanks for all the work!

**Comment by pearj** [comment]
@Lehtux we just fixed another bug that creates problems if you have maxTestSessions low. Essentially it shutdowns containers just as the tests start. You probably want to rebuild again.

**Comment by Lehtux** [comment]
@pearj @dspasojevic , this looks good! Didn't get any errors even if I had maxTestSessions low as 4. Nice job!

**Comment by peerster** [comment]
We also haven't seen this issue using your changes @pearj @dspasojevic

**Comment by pearj** [comment]
Should be fixed by #537.

Reopen if it still happens.
